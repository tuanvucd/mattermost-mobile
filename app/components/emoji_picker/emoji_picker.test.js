// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import Preferences from 'mattermost-redux/constants/preferences';

import {shallowWithIntl} from 'test/intl-test-helper';
import EmojiPicker, {filterEmojiSearchInput} from './emoji_picker.js';

describe('components/emoji_picker/EmojiPicker', () => {
    const baseProps = {
        actions: {
            getCustomEmojis: jest.fn(),
            incrementEmojiPickerPage: jest.fn(),
            searchCustomEmojis: jest.fn(),
        },
        customEmojisEnabled: false,
        customEmojiPage: 200,
        deviceWidth: 400,
        emojis: [],
        emojisBySection: [],
        fuse: {},
        isLandscape: false,
        theme: Preferences.THEMES.default,
    };

    const testCases = [
        {input: 'smile', output: 'smile'},
        {input: 'SMILE', output: 'smile'},
        {input: ':smile', output: 'smile'},
        {input: ':SMILE', output: 'smile'},
        {input: 'smile:', output: 'smile'},
        {input: 'SMILE:', output: 'smile'},
        {input: ':smile:', output: 'smile'},
        {input: ':SMILE:', output: 'smile'},
    ];

    testCases.forEach((testCase) => {
        test(`'${testCase.input}' should return '${testCase.output}'`, () => {
            expect(filterEmojiSearchInput(testCase.input)).toEqual(testCase.output);
        });
    });

    test('should match snapshot', () => {
        const wrapper = shallowWithIntl(<EmojiPicker {...baseProps}/>);
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should set rebuildEmojis to true when deviceWidth changes', () => {
        const wrapper = shallowWithIntl(<EmojiPicker {...baseProps}/>);
        const instance = wrapper.instance();

        expect(instance.rebuildEmojis).toBe(undefined);

        const newDeviceWidth = baseProps.deviceWidth * 2;
        wrapper.setProps({deviceWidth: newDeviceWidth});

        expect(instance.rebuildEmojis).toBe(true);
    });

    test('should set rebuildEmojis to true and new deviceWidth when emojis change', () => {
        const wrapper = shallowWithIntl(<EmojiPicker {...baseProps}/>);
        const instance = wrapper.instance();

        expect(instance.rebuildEmojis).toBe(undefined);
        expect(instance.deviceWidth).toBe(undefined);

        const newDeviceWidth = baseProps.deviceWidth * 2;
        const newEmojis = [{}];
        wrapper.setProps({deviceWidth: newDeviceWidth, emojis: newEmojis});

        expect(instance.rebuildEmojis).toBe(true);
        expect(instance.deviceWidth).toBe(newDeviceWidth);
    });

    test('should set rebuilt emojis when rebuildEmojis is true and searchBarAnimationComplete is true', () => {
        const wrapper = shallowWithIntl(<EmojiPicker {...baseProps}/>);
        const instance = wrapper.instance();
        instance.setState = jest.fn();
        instance.renderableEmojis = jest.spyOn(instance, 'renderableEmojis');

        instance.rebuildEmojis = true;
        const searchBarAnimationComplete = true;
        const setRebuiltEmojis = jest.spyOn(instance, 'setRebuiltEmojis');
        setRebuiltEmojis(searchBarAnimationComplete);

        expect(instance.setState).toHaveBeenCalledWith({emojis: []});
        expect(instance.rebuildEmojis).toBe(false);
    });

    test('should not set rebuilt emojis when rebuildEmojis is false and searchBarAnimationComplete is true', () => {
        const wrapper = shallowWithIntl(<EmojiPicker {...baseProps}/>);
        const instance = wrapper.instance();
        instance.setState = jest.fn();

        instance.rebuildEmojis = false;
        const searchBarAnimationComplete = true;
        const setRebuiltEmojis = jest.spyOn(instance, 'setRebuiltEmojis');
        setRebuiltEmojis(searchBarAnimationComplete);

        expect(instance.setState).not.toHaveBeenCalled();
    });

    test('should not set rebuilt emojis when rebuildEmojis is true and searchBarAnimationComplete is false', () => {
        const wrapper = shallowWithIntl(<EmojiPicker {...baseProps}/>);
        const instance = wrapper.instance();
        instance.setState = jest.fn();

        instance.rebuildEmojis = true;
        const searchBarAnimationComplete = false;
        const setRebuiltEmojis = jest.spyOn(instance, 'setRebuiltEmojis');
        setRebuiltEmojis(searchBarAnimationComplete);

        expect(instance.setState).not.toHaveBeenCalled();
    });
});
