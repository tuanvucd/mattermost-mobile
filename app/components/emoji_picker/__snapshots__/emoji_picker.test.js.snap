// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/emoji_picker/EmojiPicker should match snapshot 1`] = `
<Connect(SafeAreaIos)
  excludeHeader={true}
>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={107}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(61,60,64,0.2)",
            "paddingVertical": 5,
          },
          null,
        ]
      }
    >
      <SearchBarIos
        autoCapitalize="none"
        backgroundColor="transparent"
        blurOnSubmit={true}
        cancelTitle="Cancel"
        inputHeight={33}
        inputStyle={
          Object {
            "backgroundColor": "#ffffff",
            "color": "#3d3c40",
            "fontSize": 13,
          }
        }
        keyboardAppearance="light"
        leftComponent={null}
        onAnimationComplete={[Function]}
        onBlur={[Function]}
        onCancelButtonPress={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSearchButtonPress={[Function]}
        onSelectionChange={[Function]}
        placeholder="Search"
        placeholderTextColor="rgba(61,60,64,0.5)"
        searchIconCollapsedMargin={10}
        searchIconExpandedMargin={10}
        tintColorDelete="rgba(61,60,64,0.5)"
        tintColorSearch="rgba(61,60,64,0.8)"
        titleCancelColor="#3d3c40"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flex": 1,
        }
      }
    >
      <SectionList
        ListFooterComponent={[Function]}
        data={Array []}
        disableVirtualization={false}
        getItemLayout={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        keyboardShouldPersistTaps="always"
        maxToRenderPerBatch={10}
        onEndReached={[Function]}
        onEndReachedThreshold={0}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollToIndexFailed={[Function]}
        pageSize={30}
        removeClippedSubviews={true}
        renderItem={[Function]}
        renderSectionHeader={[Function]}
        scrollEventThrottle={50}
        sections={Array []}
        showsVerticalScrollIndicator={false}
        stickySectionHeadersEnabled={true}
        style={
          Array [
            Object {
              "backgroundColor": "#ffffff",
              "marginBottom": 35,
            },
            Object {
              "width": 370,
            },
          ]
        }
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />
      <View
        style={
          Object {
            "bottom": 0,
            "height": 35,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(61,60,64,0.1)",
              "borderTopColor": "rgba(61,60,64,0.3)",
              "borderTopWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        />
      </View>
    </View>
  </KeyboardAvoidingView>
</Connect(SafeAreaIos)>
`;
